### PR TITLE
drivers: spi: andes_atcspi200: Support dma mode

### PR DESCRIPTION
**Pull request by kevinwang821020**
1. Support the dma mode for andes_atcspi200 and use board adp_xc7k_ae350 for testing.
2. Refine the function mechanism.

### CODE REVIEW

**Comment by kevinwang821020** [comment]
Sorry to bother you reviewers, I would like to know if there are any  parts of source code that I can improve/refined.

Thanks for helping us.

File: drivers/spi/spi_andes_atcspi200.c [hunk 1]
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
...
+		data->dma_rx.dma_cfg.block_count = 1;
+		data->dma_rx.dma_cfg.dma_callback = dma_rx_callback;
+	}
+
+	/* pass our client origin to the dma: data->dma_rx.channel */
+	ret = dma_config(data->dma_rx.dma_dev, data->dma_rx.channel,
+			&data->dma_rx.dma_cfg);
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		return ret;
```
Comment (teburd) [inline drivers/spi/spi_andes_atcspi200.c:506]: Should this free the allocations on failure here?

File: drivers/spi/spi_andes_atcspi200.c [hunk 2]
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
...
+{
+	struct spi_atcspi200_data *data = user_data;
+	const struct device *spi_dev = data->self_dev;
+	struct spi_context *ctx = &data->ctx;
+
+	dma_stop(data->dma_rx.dma_dev, data->dma_rx.channel);
+	spi_rx_dma_disable(spi_dev);
+
+	if (spi_context_rx_on(ctx)) {
+		spi_dma_rx_load(spi_dev);
```
Comment (teburd) [inline drivers/spi/spi_andes_atcspi200.c:260]: the error return isn't caught here, (void)spi_dma_rx_load(spi_dev) to signal the return is ignored perhaps

File: drivers/spi/spi_andes_atcspi200.c [hunk 3]
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
...
+		spi_rx_dma_enable(spi_dev);
+		dma_start(data->dma_rx.dma_dev, data->dma_rx.channel);
```
Comment (teburd) [inline drivers/spi/spi_andes_atcspi200.c:262]: dma_start is fallible as well

File: drivers/spi/spi_andes_atcspi200.c [hunk 4]
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
...
+{
+	struct spi_atcspi200_data *data = user_data;
+	const struct device *spi_dev = data->self_dev;
+	struct spi_context *ctx = &data->ctx;
+
+	dma_stop(data->dma_tx.dma_dev, data->dma_tx.channel);
+	spi_tx_dma_disable(spi_dev);
+
+	if (spi_context_tx_on(ctx)) {
+		spi_dma_tx_load(spi_dev);
```
Comment (teburd) [inline drivers/spi/spi_andes_atcspi200.c:277]: fallible and the error is ignored, probably needs at least an assert or a (void) prefix

**Review by teburd — CHANGES_REQUESTED** [review]
Please make sure you run this against the spi test suite in some manner, there are many edge cases with the spi_buf_set that can be given to this API, would be nice to see a post here showing the test passing with this driver on hardware. It's not a requirement from me, but would give a good starting point showing the driver worked as of this point in time.

**Comment by ycsin** [comment]
@kevinwang821020 any update on this one?

File: drivers/spi/spi_andes_atcspi200.c [hunk 5]
```diff
@@ -12,14 +12,39 @@
...
 
+#ifdef CONFIG_ANDES_SPI_DMA_MODE
+
+#define ANDES_SPI_DMA_ERROR_FLAG 0x01
+#define ANDES_SPI_DMA_RX_DONE_FLAG 0x02
+#define ANDES_SPI_DMA_TX_DONE_FLAG 0x04
+#define ANDES_SPI_DMA_DONE_FLAG	\
+	(ANDES_SPI_DMA_RX_DONE_FLAG | ANDES_SPI_DMA_TX_DONE_FLAG)
+
+K_HEAP_DEFINE(chain_block, 2048);
```
[the remark this answers is not in the document]
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: Sorry for replying so late. Due to the modification for the request, there are some bug issue, so we took a while to fixed then.
We have update the source code, thanks for your suggestion.

File: drivers/spi/spi_andes_atcspi200.c [hunk 6]
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
...
+		data->dma_tx.dma_cfg.block_count = 1;
+		data->dma_tx.dma_cfg.dma_callback = dma_tx_callback;
+	}
+
+	/* pass our client origin to the dma: data->dma_tx.dma_channel */
+	ret = dma_config(data->dma_tx.dma_dev, data->dma_tx.channel,
+			&data->dma_tx.dma_cfg);
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		return ret;
```
[the remark this answers is not in the document]
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c:415]: Thank your suggestion, we did not notice that.
After the modification for previous request, this issue has been improved.

File: drivers/spi/spi_andes_atcspi200.c (hunk 4 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: Thanks for your suggestion. 
We have add some function failed detection for dma_start.

File: drivers/spi/Kconfig.andes_atcspi200 [hunk 7]
```diff
@@ -11,3 +11,12 @@ config SPI_ANDES_ATCSPI200
 	depends on DT_HAS_ANDESTECH_ATCSPI200_ENABLED
 	help
 	  Enable driver for Andes ATCSPI200 SPI controller
+
+if SPI_ANDES_ATCSPI200
+
+	config ANDES_SPI_DMA_MODE
```
Comment (ycsin) [inline drivers/spi/Kconfig.andes_atcspi200:17]:
```suggestion
config ANDES_SPI_DMA_MODE
```

File: drivers/spi/spi_andes_atcspi200.c [hunk 8]
```diff
@@ -162,31 +191,414 @@ static int configure(const struct device *dev,
...
+		error = dma_start(data->dma_tx.dma_dev, data->dma_tx.channel);
+		__ASSERT(error == 0, "dma_start was failed in tx callback");
+	}
+}
+
+/*
+ * dummy value used for transferring NOP when tx buf is null
+ * and use as dummy sink for when rx buf is null
+ */
+uint32_t dummy_rx_tx_buffer;
```
Comment (ycsin) [inline drivers/spi/spi_andes_atcspi200.c:300]: will this work for multi instances? can we just move this into device data / runtime stack?

File: drivers/spi/spi_andes_atcspi200.c [hunk 9]
```diff
@@ -195,23 +607,58 @@ static int transceive(const struct device *dev,
...
+			size_t tx_len = spi_context_total_tx_len(ctx);
+
+			chunk_len = MIN(rx_len, tx_len);
+		}
+
+		data->chunk_len = chunk_len;
 
+#ifdef CONFIG_ANDES_SPI_DMA_MODE
+		if ((data->dma_tx.dma_dev != NULL)
+		 && (data->dma_rx.dma_dev != NULL)) {
```
Comment (ycsin) [inline drivers/spi/spi_andes_atcspi200.c:644]: the new column limit is 100, this should fit:
```suggestion
		if ((data->dma_tx.dma_dev != NULL) && (data->dma_rx.dma_dev != NULL)) {
```

File: drivers/spi/spi_andes_atcspi200.c [hunk 10]
```diff
@@ -368,28 +829,104 @@ static void spi_atcspi200_irq_handler(void *arg)
...
-		sys_write32(INTST_END_INT_MSK, SPI_INTST(dev));
+		sys_write32(INTST_END_INT_MSK, SPI_INTST(cfg->base));
 
 		/* Disable all SPI interrupts */
-		sys_write32(0, SPI_INTEN(dev));
+		sys_write32(0, SPI_INTEN(cfg->base));
+
+#ifdef CONFIG_ANDES_SPI_DMA_MODE
+		if ((data->dma_tx.dma_dev != NULL)
+		 && data->dma_tx.dma_cfg.source_chaining_en) {
```
Comment (ycsin) [inline drivers/spi/spi_andes_atcspi200.c:850]:
```suggestion
		if ((data->dma_tx.dma_dev != NULL) && data->dma_tx.dma_cfg.source_chaining_en) {
```

File: drivers/spi/spi_andes_atcspi200.c [hunk 11]
```diff
@@ -368,28 +829,104 @@ static void spi_atcspi200_irq_handler(void *arg)
...
+
+			spi_tx_dma_disable(dev);
+			dma_stop(data->dma_tx.dma_dev, data->dma_tx.channel);
+			data->dma_tx.block_idx = 0;
+			data->dma_tx.dma_blk_cfg.next_block = NULL;
+		}
+
+		if ((data->dma_rx.dma_dev != NULL)
+		 && data->dma_rx.dma_cfg.source_chaining_en) {
```
Comment (ycsin) [inline drivers/spi/spi_andes_atcspi200.c:859]:
```suggestion
		if ((data->dma_rx.dma_dev != NULL) && data->dma_rx.dma_cfg.source_chaining_en) {
```

**Review by ycsin — DISMISSED** [review]

**Comment by nashif** [comment]
@kevinwang821020 can you please review?

**Comment by kevinwang821020** [comment]
> Please make sure you run this against the spi test suite in some manner, there are many edge cases with the spi_buf_set that can be given to this API, would be nice to see a post here showing the test passing with this driver on hardware. It's not a requirement from me, but would give a good starting point showing the driver worked as of this point in time.

@teburd ,
Thanks for your suggestion.
Yes, there are lots of edge cases, therefore, it took us a while to improve and verify it. Also, the dma mode(especially with chain transfer) takes more time.

Below is the test passing photo, it's with dma chain thransfer mode.

image

File: drivers/spi/spi_andes_atcspi200.c (hunk 8 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c:300]: Yes, it work for multi instances, the data send/to this variable is noused.
I think move into runtime stack may cause some problem, because after the spi_dma_tx/rx_load function has return, the variable will be release, after that the address may be used by other function. But at that time, the transfer may still ongoing, so the variable used by other function in the address will be polluted.

File: drivers/spi/spi_andes_atcspi200.c (hunk 11 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: Thanks for your suggestion!
We have update the source code.

**Comment by kevinwang821020** [comment]
Sorry to bother you reviewers, we would like to know if there are any suggestion or request for this PR.

Thanks for helping us.

**Comment by ycsin** [comment]
Looks good to me, I think it would be better if the SPI ztest results are pasted in the OP.

cc @teburd could you take another look at this please?

File: drivers/spi/Kconfig.andes_atcspi200 [hunk 12]
```diff
@@ -11,3 +11,12 @@ config SPI_ANDES_ATCSPI200
 	depends on DT_HAS_ANDESTECH_ATCSPI200_ENABLED
 	help
 	  Enable driver for Andes ATCSPI200 SPI controller
+
```
Comment (tbursztyka) [inline drivers/spi/Kconfig.andes_atcspi200:14]: change your commit title to: "drivers: spi: Support dma mode for atcspi200"

File: drivers/spi/spi_andes_atcspi200.c [hunk 13]
```diff
@@ -162,31 +191,414 @@ static int configure(const struct device *dev,
...
+static int spi_transfer_dma(const struct device *dev)
+{
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	struct spi_atcspi200_data * const data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	uint32_t data_len, tctrl, dma_rx_enable, dma_tx_enable;
+	int error = 0;
+
+	data_len = data->chunk_len - 1;
+
```
Comment (tbursztyka) [inline drivers/spi/spi_andes_atcspi200.c:543]: you can remove that empty line as your are testing in the if below the variable you assigned previously (apply that were relevant)

File: drivers/spi/spi_andes_atcspi200.c [hunk 14]
```diff
@@ -195,23 +607,57 @@ static int transceive(const struct device *dev,
...
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
 	struct spi_atcspi200_data * const data = dev->data;
 	struct spi_context *ctx = &data->ctx;
 	int error, dfs;
+	size_t chunk_len;
 
 	spi_context_lock(ctx, asynchronous, cb, userdata, config);
-
 	error = configure(dev, config);
+
```
Comment (tbursztyka) [inline drivers/spi/spi_andes_atcspi200.c:618]: like here for instance (this change was unnecessary)

File: drivers/spi/spi_andes_atcspi200.h [hunk 15]
```diff
@@ -11,34 +11,21 @@ LOG_MODULE_REGISTER(spi_atcspi200);
...
-#define REG_DATA        0x2c
-#define REG_CTRL        0x30
-#define REG_STAT        0x34
-#define REG_INTEN       0x38
-#define REG_INTST       0x3c
-#define REG_TIMIN       0x40
-#define REG_MCTRL       0x50
-#define REG_SLVST       0x60
-#define REG_SDCNT       0x64
-#define REG_CONFIG      0x7c
```
Comment (tbursztyka) [inline drivers/spi/spi_andes_atcspi200.h:29]: keel these macros

File: drivers/spi/spi_andes_atcspi200.h [hunk 16]
```diff
@@ -11,34 +11,21 @@ LOG_MODULE_REGISTER(spi_atcspi200);
...
-#define SPI_INTEN(dev)	(SPI_BASE + REG_INTEN)
-#define SPI_INTST(dev)	(SPI_BASE + REG_INTST)
-#define SPI_TIMIN(dev)	(SPI_BASE + REG_TIMIN)
-#define SPI_CONFIG(dev)	(SPI_BASE + REG_CONFIG)
+#ifdef CONFIG_ANDES_SPI_DMA_MODE
+#include <zephyr/dt-bindings/dma/andes_dma.h>
+#include <zephyr/drivers/dma.h>
+#endif
+
+#define SPI_TFMAT(base)		(base + 0x10)
```
Comment (tbursztyka) [inline drivers/spi/spi_andes_atcspi200.h:19]: and keep using them in these macros.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Don't mix changes that are unrelated to the commit. It's better to split into proper commits then.

File: drivers/spi/spi_andes_atcspi200.c (hunk 13 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c:543]: Ok, thanks a lot.

File: drivers/spi/Kconfig.andes_atcspi200 (hunk 12 above)
Comment (kevinwang821020) [reply drivers/spi/Kconfig.andes_atcspi200:14]: OK, I got it.
Thank for your recommendation, we had done the modification.

File: drivers/spi/spi_andes_atcspi200.h (hunk 16 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.h:19]: OK, no problem.
We have update our source code.
Thank you!

**Comment by kevinwang821020** [comment]
Sorry for bothering the reviewers, we know that the version 3.5 is about to be released, and we look forward that this source code can be included in this release. Is there any good way to do this or any recommendations?

Thank you.

**Comment by kevinwang821020** [comment]
Due to the PR #59860(for dma driver) was merged, we rebase the source code for that.
Now with the CONFIG_DMA enable, the spi can do the transfer via dma on board adp_xc7k_ae350.

File: drivers/spi/Kconfig.andes_atcspi200 (hunk 12 above)
Comment (tbursztyka) [reply drivers/spi/Kconfig.andes_atcspi200:14]: I still read "drivers: spi: Support dma mode"

**Review by tbursztyka — COMMENTED** [review]
You have updated your source, but not pushed force to github then?

File: drivers/spi/spi_andes_atcspi200.c [hunk 17]
```diff
@@ -162,31 +191,414 @@ static int configure(const struct device *dev,
...
+	/* give the client dev as arg, as the callback comes from the dma */
+	data->dma_tx.dma_cfg.user_data = data;
+
+	if (data->dma_tx.dma_cfg.source_chaining_en) {
+
+		data->dma_tx.dma_cfg.block_count = ctx->tx_count;
+		data->dma_tx.dma_cfg.dma_callback = NULL;
+		data->dma_tx.block_idx = 0;
+		struct dma_block_config *blk_cfg = &data->dma_tx.dma_blk_cfg;
+		const struct spi_buf *current_tx = ctx->current_tx;
```
[the remark this answers is not in the document]
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c:358]: @tbursztyka  so sorry for that, I have pushed the correct commit before, but when I do the rebase for PR #59860 last week, it seems like i do some incorrect rebase operation, so those commit was disappered.
I have corrected the git branch, it's my mistake.

Also I have update the source code for your new suggestion.

Thanks a lot!

File: drivers/spi/Kconfig.andes_atcspi200 [hunk 18]
```diff
@@ -11,3 +11,12 @@ config SPI_ANDES_ATCSPI200
...
+if SPI_ANDES_ATCSPI200
+
+config ANDES_SPI_DMA_MODE
+	bool "Using DMA mode for spi"
+	default y
+	depends on DMA
+
+endif # SPI_ANDES_ATCSPI200
```
Comment (tbursztyka) [inline drivers/spi/Kconfig.andes_atcspi200:22]: that should go in the second commit

Comment (kevinwang821020) [reply drivers/spi/Kconfig.andes_atcspi200:22]: OK I got it

**Comment by kevinwang821020** [comment]
> @kevinwang821020 any update on this one?

Hi @ycsin @teburd
Sorry for  bothering you.
I have update the source for your request before, and I just want to confirm  if my changes meet your requirements.

File: drivers/spi/spi_andes_atcspi200.c [hunk 19]
```diff
@@ -162,31 +191,412 @@ static int configure(const struct device *dev,
...
+		}
+	}
+
+	return 0;
+}
+
+static inline void dma_rx_callback(const struct device *dev, void *user_data,
+				uint32_t channel, int status)
+{
+	struct spi_atcspi200_data *data = user_data;
```
Comment (teburd) [inline drivers/spi/spi_andes_atcspi200.c:257]: Why not use the device itself as the user_data for dma here instead if you need to get the struct device?

Can drop the self_dev pointer in data then

File: drivers/spi/spi_andes_atcspi200.c [hunk 20]
```diff
@@ -12,14 +12,39 @@
...
+	struct dma_block_config dma_blk_cfg;
+	struct dma_block_config chain_block[MAX_CHAIN_SIZE];
+	uint8_t priority;
+	bool src_addr_increment;
+	bool dst_addr_increment;
+};
+#endif
+
 struct spi_atcspi200_data {
+	const struct device *self_dev;
```
Comment (teburd) [inline drivers/spi/spi_andes_atcspi200.c:37]: This shouldn't be needed

**Review by teburd — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_andes_atcspi200.c (hunk 20 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: OK, yes, that is more appropriate.
I have update the source code.

Thank you.

**Review by teburd — DISMISSED** [review]
I'm ok with it

File: drivers/spi/spi_andes_atcspi200.c [hunk 21]
```diff
@@ -398,43 +401,41 @@ static void spi_atcspi200_irq_handler(void *arg)
...
-										\
-		IRQ_CONNECT(DT_INST_IRQN(n),					\
-			    DT_INST_IRQ(n, priority),				\
-			    spi_atcspi200_irq_handler,				\
-			    DEVICE_DT_INST_GET(n),				\
-			    0);							\
-										\
-	}
+#define SPI_INIT(n)							\
+	static struct spi_atcspi200_data spi_atcspi200_dev_data_##n = {\
```
Comment (ycsin) [inline drivers/spi/spi_andes_atcspi200.c:938]:
```suggestion
	static struct spi_atcspi200_data spi_atcspi200_dev_data_##n = { \
```

**Review by ycsin — COMMENTED** [review]
nit

File: drivers/spi/spi_andes_atcspi200.c (hunk 21 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: Thanks, that is great!
I update the source code and rebase it.

**Review by ycsin — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]
